config/initializers/synth.rb

```diff
@@ -0,0 +1,5 @@
+Rails.application.config.to_prepare do
+  Provider::Synth.configure do |config|
+    config.api_key = ENV["SYNTH_API_KEY"]
+  end
+end
```

Comment:
The issue I think we'll run into here is for self-hosters.  They'll need to be able to configure their Synth API key directly in the UI via the [settings model](https://github.com/maybe-finance/maybe/blob/main/app/models/setting.rb), which we haven't configured quite yet, but will need to soon.

You had mentioned that a big reason for this refactor is due to `initialized?` always returning false during the tests.  I'm thinking that may be more of a symptom of tightly-coupled dependencies rather than a reason to change the class itself.  I'm not sure which tests were failing on you, but I think we should try to address it elsewhere so we can keep the config for `Synth` available at runtime.  Let me know which tests are causing problems and I'd be happy to throw around some ideas.

Something that may help out is the `with_env_override` helper, which allows us to toggle environment variables on and off during tests for special cases where we need to test "enabled/disabled" behaviors:

https://github.com/maybe-finance/maybe/blob/457247da8e40254c469a1133fefbba4ffc5013b5/test/test_helper.rb#L57-L59

Here's an example usage:

https://github.com/maybe-finance/maybe/blob/457247da8e40254c469a1133fefbba4ffc5013b5/test/controllers/registrations_controller_test.rb#L28

Reply:
It was specificaly the `sync` tests were failing because of the new early return we discuss in the other thread. If we were to remove the early return and instead proceed with the sync generating possibly invalid balances, then that would also remove the need for this refactor

Reply:
Yes this does allow for configuration at runtime, but it would be more of a "re-configuration" (initializes, then user changes the api key in their settings), which feels like an extra layer of abstraction (global state) that is not necessary.

Here is the [original data providers PR](https://github.com/maybe-finance/maybe/pull/574) where I had started down this "global state" path and shows how we ended up with the current design which is more tightly integrated with the domain models.

I think rather than focusing on configuration and the `initialized?` method that checks whether an API key is present (which still doesn't guarantee rates are present), we need to dig into our calculator class and refactor the logic to make these decisions based on the presence/absence of the exchange rates as I mentioned in the other comment.

In other words, if we're syncing foreign currency account XYZ, we need to figure out which exchange rates are needed, check if we have all of them available (regardless of providers as they could be in our database), and then proceed/abort the sync depending on that response.

Reply:
And then for the UI warnings/errors, we can check on an account-by-account basis by populating these during the sync process (these are currently not in use, but I had added as placeholders):

https://github.com/maybe-finance/maybe/blob/6e59fdb369d18dcd1163e9088f272a716ace6acb/db/schema.rb#L93-L94

That way, we can simply say:

```erb
<% if @account.sync_errors.present? %>
  <%= render partial: "accounts/some-partial", locals: { errors: @account.sync_errors } %>
<% end %>
```